fsspec/implementations/dropboxdrivefs.py

```diff
@@ -0,0 +1,244 @@
+
+        This is only called when position is still at zero,
+        and read() is called without a byte-count.
+        """
+        if not isinstance(self.cache, AllBytes):
+            r = self.session.get(self.url, **self.kwargs)
+            r.raise_for_status()
+            out = r.content
+            self.cache = AllBytes(out)
+            self.size = len(out)
+
+    def _fetch_range(self, start, end):
+        """Download a block of data
+
+        The expectation is that the server returns only the requested bytes,
+        with HTTP code 206. If this is not the case, we first check the headers,
+        and then stream the output - if the data size is bigger than we
+        requested, an exception is raised.
+        """
+        kwargs = self.kwargs.copy()
+        headers = kwargs.pop("headers", {})
+        headers["Range"] = "bytes=%i-%i" % (start, end - 1)
+        r = self.session.get(self.url, headers=headers, stream=True, **kwargs)
+        if r.status_code == 416:
+            # range request outside file
+            return b""
+        r.raise_for_status()
+        if r.status_code == 206:
+            # partial content, as expected
+            out = r.content
+        elif "Content-Length" in r.headers:
+            cl = int(r.headers["Content-Length"])
+            if cl <= end - start:
+                # data size OK
+                out = r.content
+            else:
+                raise ValueError(
+                    "Got more bytes (%i) than requested (%i)" % (cl, end - start)
+                )
+        else:
+            cl = 0
+            out = []
+            for chunk in r.iter_content(chunk_size=2 ** 20):
+                # data size unknown, let's see if it goes too big
+                if chunk:
+                    out.append(chunk)
+                    cl += len(chunk)
+                    if cl > end - start:
+                        raise ValueError(
+                            "Got more bytes so far (>%i) than requested (%i)"
+                            % (cl, end - start)
+                        )
+                else:
+                    break
+            out = b"".join(out)
+        return out
+
+    def _upload_chunk(self, final=False):
+        self.cursor.offset += self.buffer.seek(0, 2)
+        if final:
+            self.dbx.files_upload_session_finish(
+                self.buffer.getvalue(), self.cursor, self.commit
+            )
+        else:
+            self.dbx.files_upload_session_append(
+                self.buffer.getvalue(), self.cursor.session_id, self.cursor.offset
+            )
+
+    def _initiate_upload(self):
+        """ Initiate the upload session
+        """
+        session = self.dbx.files_upload_session_start(self.buffer.getvalue())
+        self.commit = dropbox.files.CommitInfo(path=self.path)
+        self.cursor = dropbox.files.UploadSessionCursor(
+            session_id=session.session_id, offset=self.offset
+        )
+
+
+class AllBytes(object):
```

Comment:
Most likely should implement this in `fsspec.caching` (or move the one from fsspec.implementations.http there).